### PR TITLE
Fix memory issues / create raw geometry type class

### PR DESCRIPTION
**Pull request by ewestern**
The goal for this PR is to allow the type system to take care of as much of the memory issues as possible. This means:
Make many raw geometry functions use a "Geometry" type class that has implementations for objects that both have and do not have finalizers associated with them.

As it stands, all of the tests work without memory issues (though one test is incorrect).

There are still some things that need cleaning up. 
Most importantly, the test suite needs to be greatly expanded. Specifically, we need tests for convertGeometryFromRaw and convertGeometryToRaw for *each* of the geometry types.
Additionally, none of the secondary modules (Topology, Prepared, Buffer) are tested.

### CODE REVIEW

File: tests/Main.hs [hunk 1]
```diff
@@ -82,17 +87,47 @@ main = hspec $ do
...
       let lsg :: Geometry LineString = withSomeGeometry (readHex linestringBS) $ \case
                   ls@(LineStringGeometry _ _) -> ls
                   _ -> error "asda"
       lsg `shouldBe` linestring
+    it "Parse a multipolygon" $ do
+      pending
+      let multig :: Geometry MultiPolygon = withSomeGeometry (readHex multiPolygonStringBS) $ \case
+                  g@(MultiPolygonGeometry _ _) -> g
+                  _ -> error "asda"
+      (show multig) `shouldBe` (show linestring)
```
Comment (ewestern) [inline tests/Main.hs]: @jamiecook Multipolygon and Linestring have different show instances. We need to compare the results with an actual MultiPolygon object.

File: geos.cabal [hunk 2]
```diff
@@ -36,7 +36,7 @@ Library
   build-tools: hsc2hs
   exposed-modules:
       Data.Geometry.Geos.Topology
-    , Data.Geometry.Geos.Geometry
+      Data.Geometry.Geos.Geometry
```
Comment (jamiecook) [inline geos.cabal]: i think this might be a typo

File: tests/Main.hs (hunk 1 above)
Comment (jamiecook) [reply tests/Main.hs]: Right you are sir... how about this https://github.com/ewestern/geos/pull/8

**Comment by jamiecook** [comment]
Peter, I've pushed a bit more into my test branch (based off this branch) 

I've made a start on some basic testing for geo operations. 
I started with contains (because it's what I need first and one of the easiest to reason about) and I've been able to reproduce the following memory error: 

image

As you can see it happens only for MultiPolygons.

**Comment by jamiecook** [comment]
Thanks for the merge, I've been trying to expose some more memory problems by running through the problem that we have here (filtering some points with some multipolygons read from a PostGIS database) and have so far been unable to get it to crash. (except that some of the multipolygons aren't actually multi at all).

File: src/Data/Geometry/Geos/Geometry.hs [hunk 3]
```diff
@@ -125,8 +127,8 @@ convertLineStringToRaw (LineString cs) s = do
 
 convertPolygonToRaw :: R.Geometry a => Polygon -> SRID -> Geos a
 convertPolygonToRaw (Polygon lrs) s = do
-  ext <- convertLinearRingToRaw (V.head lrs) s
-  inn <- (flip convertLinearRingToRaw $ s) `V.mapM` V.tail lrs
+  ext :: R.GeomConst <- convertLinearRingToRaw (V.head lrs) s
+  inn :: V.Vector R.GeomConst <- (flip convertLinearRingToRaw $ s) `V.mapM` V.tail lrs
   R.createPolygon ext (V.toList inn)  >>= R.setSRID s
 
 convertMultiPointToRaw :: R.Geometry a => MultiPoint -> SRID -> Geos a
```
Comment (jamiecook) [inline src/Data/Geometry/Geos/Geometry.hs]: Would I be right in thinking that these other converters also need the GeomConst treatment to stop the sub-objects (Points) of the parent (MultiPoint) trying to free themselves after the parent has already destroyed/free'd their underlying pointers?

Or have I misunderstood (I estimate that at > 50% probability)

Comment (ewestern) [reply src/Data/Geometry/Geos/Geometry.hs]: That's right (though it's sufficient to ensure that the R.create* functions take a GeomConst so I'm removing the scoped type variables here.
Interestingly, I was unable to reproduce the same memory bug with the MultiPoint, MultiLineString, (etc) functions.

**Pull request merged**